Cards/CardContainer.gd

```diff
@@ -10,93 +10,184 @@ class_name CardContainer
+
+
+func is_card_queued() -> bool:
+	return queued_card != null
+
+
+func deal_to_hand_limit() -> void:
+	# deal to our starting hand size, ignoring amount of cards we already have
+	var amount_to_draw: int = maxi(0, starting_hand_size - cards_in_hand.size())
+	draw_cards(amount_to_draw)
+
+
+func draw_cards(amount: int) -> void:
+	for card_index: int in amount:
+		# limit amount of cards to a maximum
+		if cards_in_hand.size() < max_hand_size:
+			_draw_card()
 
 
 func discard_all_cards() -> void:
-	for card: CardWorld in cards:
-		card.queue_free()
-	cards.clear()
+	while cards_in_hand.size() > 0:
+		_discard_last_card()
 
 
-func remove_card(card: CardWorld) -> void:
-	var card_index: int = cards.find(card)
-	cards[card_index].queue_free()
-	cards.remove_at(card_index)
+func discard_random_card(amount: int) -> void:
+	var valid_cards_to_discard: Array[CardWorld] = []
+	
+	for card in cards_in_hand:
+		if card != queued_card:
+			valid_cards_to_discard.append(card)
+	
+	var amount_to_discard: int = mini(valid_cards_to_discard.size(), amount)
+	for discarding_index: int in amount_to_discard:
+		var random_index: int = Helpers.get_random_array_index(valid_cards_to_discard)
+		discard_card(valid_cards_to_discard[random_index])
+		valid_cards_to_discard.remove_at(random_index)
+
+
+func discard_card(card: CardWorld) -> void:
+	var card_index: int = cards_in_hand.find(card)
+	_discard_card_at_index(card_index)
+
+
+func get_draw_pile_size() -> int:
+	return draw_pile.size()
+
+
+func get_discard_pile_size() -> int:
+	return discard_pile.size()
+
+
+func _draw_card() -> void:
+	# if draw pile is empty, shuffle discard pile into draw pile. Clear discard pile.
+	if draw_pile.size() <= 0:
+		# every card is in our hand. Exit because there's nothing to draw
+		if discard_pile.size() <= 0:
+			return
+		discard_pile.shuffle()
+		draw_pile = discard_pile.duplicate()
+		discard_pile.clear()
+	
+	# top card on the draw pile
+	var drawn_card: CardBase = draw_pile[0]
+	draw_pile.remove_at(0)
+	
+	_create_card_in_world(drawn_card)
+	
+	on_card_counts_updated.emit()
+
+
+func _create_card_in_world(card_data: CardBase):
+	var card_instance: CardWorld = card_scene.instantiate()
+	add_child(card_instance)
+	cards_in_hand.append(card_instance)
+
+	card_instance.init_card(card_data)
+	
+	# bind mouse events
+	var card_click_handler: ClickHandler = card_instance.get_click_handler()
+	card_click_handler.on_click.connect(_on_card_clicked.bind(card_instance))
+	card_click_handler.on_mouse_hovering.connect(_on_card_hovering.bind(card_instance))
+	card_click_handler.on_unhover.connect(_on_card_unhovered.bind(card_instance))
+
+
+func _discard_card_at_index(card_index: int) -> void:
+	var card: CardWorld = cards_in_hand[card_index]
+	
+	# add to discard pile
+	discard_pile.append(card.card_data)
+	
+	# remove from world
+	cards_in_hand.remove_at(card_index)
+	card.queue_free()
+	
+	on_card_counts_updated.emit()
+
+
+func _discard_last_card() -> void:
```

Comment:
we should be careful of how we use this function later on. For now it's only used to discard all cards, but it could create some issues if not handled properly, since we append cards but we also remove cards from last index (we could end up removing a card we just added)

Reply:
Don't think we will use it for anything except discarding all cards unless we end up needing it elsewhere internally. Externally you shouldn't ever call this